tsdb/head_append.go

```diff
@@ -290,15 +290,22 @@ func (s *memSeries) appendable(t int64, v float64) error {
 	if t > c.maxTime {
 		return nil
 	}
-	if t < c.maxTime {
-		return storage.ErrOutOfOrderSample
-	}
+
 	// We are allowing exact duplicates as we can encounter them in valid cases
 	// like federation and erroring out at that time would be extremely noisy.
-	if math.Float64bits(s.sampleBuf[3].v) != math.Float64bits(v) {
+	if t == c.maxTime && math.Float64bits(s.sampleBuf[3].v) != math.Float64bits(v) {
 		return storage.ErrDuplicateSampleForTimestamp
 	}
-	return nil
+
+	// t < c.maxTime
+	lenBuf := len(s.sampleBuf)
+	for i := range s.sampleBuf {
+		sample := s.sampleBuf[lenBuf-i-1]
+		if math.Float64bits(sample.v) == math.Float64bits(v) && sample.t == t {
+			return nil
+		}
+	}
```

Comment:
My gut feeling: let's unroll the loop and just have 4 if statements since slice length is fixed, much simpler to read. When there is a slurry of out of order samples, it can prolly give a small edge in latency too.

Reply:
👍 will do.

i was under the impression the compiler will unroll it, since the length is known. but maybe not, since golang is more of "compile fast" club, also unrolled is more readable like you said.